### PR TITLE
In Dns.RunAsync calls AfterResolution when canceled

### LINKED ISSUE #92045: System.Net.Dns.RunAsync() does not decrement telemetry for canceled requests
If a call that ends up in `Dns.RunAsync()` is canceled, nothing ever calls `NameResolutionTelemetry.AfterResolution()` and so the variable `NameResolutionTelemetry._currentLookups` remains forever incremented by one.

The unmatched `BeforeResolution()` is called in `Dns.RunAsync()` here: https://github.com/dotnet/runtime/blob/9101b853a1f5194de3e7bc495454a57df0da300e/src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs#L652C16-L652C16

image

### PR DESCRIPTION
**Pull request by rokonec**
Fixes: #92045

## Context
When Dns.RunAsync is canceled before it runs its action, pairing `Log.AfterResolution` was not called causing permanently incorrect telemetry counters value. See description of  #92045

## Changes
Added `AfterResolution` in `OnlyOnCanceled` registered continuation.

### CODE REVIEW

File: src/libraries/System.Net.NameResolution/System.Net.NameResolution.sln [hunk 1]
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
+﻿
```
Comment (liveans) [inline src/libraries/System.Net.NameResolution/System.Net.NameResolution.sln:1]: Can you revert unintended changes in this file?
Same applies to `System.Net.Sockets.sln` as well.

Comment (rokonec) [reply src/libraries/System.Net.NameResolution/System.Net.NameResolution.sln:1]: Oh. Not again!!!! 
I will.

File: src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs [hunk 2]
```diff
@@ -694,6 +694,9 @@ private static Task<TResult> RunAsync<TResult>(Func<object, long, TResult> func,
                         lock (s_tasks)
                         {
                             ((ICollection<KeyValuePair<object, Task>>)s_tasks).Remove(new KeyValuePair<object, Task>(key!, task));
+                            // Since it was canceled, func(..) had not executed and call AfterResolution it needs to be called here.
+                            NameResolutionTelemetry.Log.AfterResolution(key!, startingTimestamp);
```
Comment (rokonec) [inline src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:698]: Should we add `new OperationCanceledException());` as `exception` argument so Log then states that particular resolution has been canceled?

Comment (MihaZupan) [reply src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:698]: Yes, this should be reported as a failure.
Nit: you shouldn't need the `!` on the key as the nullability is implied by the `!` 1 line above.

**Review by MihaZupan — APPROVED** [review]

File: src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs (hunk 2 above)
Comment (antonfirsov) [reply src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:698]: Instead of newing up a new exception, I think we can use a static instance, since it's only being used to determine `error.type` for the `dns.lookup.duration` metric.

**Review by antonfirsov — COMMENTED** [review]
What happens if the `CancellationToken` is fired during the invocation of the delegate passed to `prevTask.ContinueWith()` (not the previously queued task)? Isn't it possible that both continuation delegates will be invoked leading to double call into `AfterResolution`?

Also, can we attempt to design a test for this? Idea for a "mini stress test": have a bunch parallel calls in a separate `RemoteExecutor` process, cancel some of them, count the `ResolutionStart` & `ResolutionStop` events.

File: src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs (hunk 2 above)
Comment (rokonec) [reply src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:698]: > Yes, this should be reported as a failure. Nit: you shouldn't need the `!` on the key as the nullability is implied by the `!` 1 line above.

It is screaming at me without it. In both static analysis and actual compile.

Comment (rokonec) [reply src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs]: > Instead of newing up a new exception, I think we can use a static instance, since it's only being used to determine `error.type` for the `dns.lookup.duration` metric.

I thought about that, however from my point of view, it is not on hot path, and in cancelation path performance is rarely an issue.

**Comment by rokonec** [comment]
> What happens if the `CancellationToken` is fired during the invocation of the delegate passed to `prevTask.ContinueWith()` (not the previously queued task)? Isn't it possible that both continuation delegates will be invoked leading to double call into `AfterResolution`?

Because we dont pass CT to func() `prevTask.ContinueWith()` if the `CancellationToken` is fired during the invocation of the delegate the delegate will complete and call `AfterRFesolution` and `task.ContinueWith` for `TaskContinuationOptions.OnlyOnCanceled` will not fired as this do not care for original task CT but only if task status indicated completion - which it will not.

TLDR: I believe it is OK

> Also, can we attempt to design a test for this? Idea for a "mini stress test": have a bunch parallel calls in a separate `RemoteExecutor` process, cancel some of them, count the `ResolutionStart` & `ResolutionStop` events.

Yes it will be nice to have test to validate balanced `ResolutionStart` & `ResolutionStop`, however due to various caching on DNS I have no idea how to validate that failure case when resolution is canceled after it was added into task Continuation and before this continuation has been scheduled/run. I have tried it locally and failed to came up with stable solution. If cancelation toke is canceled before call of `GetHostAddressesAsync`, RunAsync is not called at all, and neither After or Before resolution can possibly be called. I don't see how to overcome this in RemoteExecutor. 

If this is a high concert I would recommend to handle it in isolated PR, as this requires, IMO, some refactoring and investigations.
